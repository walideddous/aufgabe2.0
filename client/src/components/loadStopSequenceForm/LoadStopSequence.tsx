import React, { useState, useCallback, useEffect } from "react";
import { AutoComplete, Card, Form, Select, Button } from "antd";
// Import types
import { TStopSequence } from "../../types/types";

interface TLoadStopSequence {
  show: boolean;
  currentMode: string[];
  stopSequenceList: TStopSequence[];
  currentStopSequence: TStopSequence | undefined;
  onStopsQuery: (modes: string[]) => void;
  onStopSequenceSearch: (name: string) => void;
  onDeleteStopSequence: (id: string) => void;
  onDisplayStopSequence: (modes: string[], key: string) => void;
}

const LoadStopSequence = ({
  show,
  currentMode,
  stopSequenceList,
  currentStopSequence,
  onStopsQuery,
  onStopSequenceSearch,
  onDeleteStopSequence,
  onDisplayStopSequence,
}: TLoadStopSequence) => {
  const [form] = Form.useForm();
  const [search, setSearch] = useState<string>("");
  const [selectValue, setSelectValue] = useState<string>(
    "verkehrsmittel typ auswählen"
  );

  // Auto complete component
  const { Option } = AutoComplete;

  useEffect(() => {
    if (currentMode[0]) {
      setSelectValue(currentMode[0] + "");
    }
  }, [currentMode]);

  // implementing the debouncing
  useEffect(() => {
    if (search) {
      const dispatchRequest = setTimeout(() => {
        onStopSequenceSearch(search);
      }, 1000);

      return () => clearTimeout(dispatchRequest);
    }
  }, [search, onStopSequenceSearch]);

  // handle the drop menu to display the choosed Modes on Map
  const handleModeChange = useCallback(
    (value: string) => {
      if (value !== "verkehrsmittel typ auswählen") {
        onStopsQuery([value + ""]);
        setSelectValue(value + "");
      }
    },
    [onStopsQuery]
  );

  const handleSelect = useCallback(
    (input: string, options) => {
      const response = stopSequenceList.filter(
        (el: any) => el.key === options.key
      )[0];
      if (response) {
        onDisplayStopSequence(response.modes, options.key);
      }
    },
    [stopSequenceList, onDisplayStopSequence]
  );

  return (
    <Card>
      <Form form={form} layout="vertical">
        {!show && (
          <Form.Item label="Verkehrsmittel typ">
            <Select
              id="mode_selector"
              value={selectValue}
              onChange={handleModeChange}
            >
              <Option value="verkehrsmittel typ auswählen" id="Modus_auswählen">
                verkehrsmittel typ auswählen
              </Option>
              <Option value="13" id="13">
                13
              </Option>
              <Option value="5" id="5">
                5
              </Option>
              <Option value="8" id="8">
                8
              </Option>
              <Option value="9" id="9">
                9
              </Option>
              <Option value="2" id="2">
                2
              </Option>
              <Option value="4" id="4">
                4
              </Option>
            </Select>
          </Form.Item>
        )}
        {show && (
          <>
            <Form.Item label="Haltestellensequenz name">
              <AutoComplete
                id="stopSequence_autoComplete"
                onChange={(input: string) => {
                  setSearch(input);
                }}
                value={search}
                onSelect={handleSelect}
                placeholder="Haltestellensequenz nach Namen suchen"
                allowClear={true}
              >
                {stopSequenceList &&
                  stopSequenceList.map((el: any) => (
                    <Option value={el.name} key={el.key}>
                      {el.name}
                    </Option>
                  ))}
              </AutoComplete>
            </Form.Item>
            {currentStopSequence && (
              <div style={{ display: "flex", justifyContent: "flex-end" }}>
                <Button
                  danger
                  id="delete_stopSequence"
                  disabled={
                    search &&
                    currentStopSequence &&
                    search === currentStopSequence.name
                      ? false
                      : true
                  }
                  onClick={() => {
                    if (currentStopSequence) {
                      const { key } = currentStopSequence;
                      onDeleteStopSequence(key);
                      setSearch("");
                    }
                  }}
                >
                  Haltestellensequenz löschen
                </Button>
              </div>
            )}
          </>
        )}
      </Form>
    </Card>
  );
};

export default React.memo(LoadStopSequence);
