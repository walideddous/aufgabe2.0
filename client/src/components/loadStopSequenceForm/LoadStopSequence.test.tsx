import React from "react";
import { mount, shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import LoadStopSequence from "./LoadStopSequence";

describe("LoadStopSequence component", () => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  let shallowWrapper: any;
  let props = {
    stopSequenceList: [
      {
        _id: "c03295ea-5a3f-43e8-83ea-7736ce82cfd9",
        name: "St. Gallen to Zürich HB",
        date: ["2020-10-16", "2020-11-27"],
        schedule: [
          {
            day: ["monday", "tuesday", "wednesday", "thursday", "friday"],
            time: [
              {
                start: "07:00",
                end: "12:00",
              },
              {
                start: "13:00",
                end: "18:00",
              },
            ],
          },
          {
            day: ["saturday", "sunday"],
            time: [
              {
                start: "07:00",
                end: "12:00",
              },
            ],
          },
        ],
        modes: "13",
        stopSequence: [
          {
            _id: "5f62045b0d5658001cd910c4",
            name: "St. Gallen",
            modes: ["13", "5"],
            coord: {
              WGS84: {
                lat: 47.42318,
                lon: 9.3699,
              },
            },
          },
          {
            _id: "5f62045b0d5658001cd910c1",
            name: "St. Gallen Bruggen",
            modes: ["13"],
            coord: {
              WGS84: {
                lat: 47.4072,
                lon: 9.32965,
              },
            },
          },
        ],
      },
    ],
    stateDND: {
      vorschlag: {
        title: "Suggestion",
        items: [],
      },
      trajekt: {
        title: "Stop sequence",
        items: [],
      },
    },
    currentStopSequence: {},
  };

  const spyOnHandleUpdateAfterSave = jest.fn();
  const spyOnOnSendRequest = jest.fn();
  const spyOnOndisplayStopSequence = jest.fn();
  const spyOnHandleDeleteStopSequence = jest.fn();
  const spyOnClearAll = jest.fn();

  beforeEach(() => {
    shallowWrapper = shallow(
      <LoadStopSequence
        {...props}
        handleUpdateAfterSave={spyOnHandleUpdateAfterSave}
        onSendRequest={spyOnOnSendRequest}
        ondisplayStopSequence={spyOnOndisplayStopSequence}
        handleDeleteStopSequence={spyOnHandleDeleteStopSequence}
        onClearAll={spyOnClearAll}
      />
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should match the LoadStopSequence component snapshot", () => {
    expect(toJSON(shallowWrapper)).toMatchSnapshot();
  });

  it("Should disptach the onSendRequest props function when we choose the mode", () => {
    const modeSelector = shallowWrapper.find("#mode_selector");

    expect(modeSelector.props().value).toEqual("Choose mode");

    modeSelector.simulate("change", "4");

    expect(spyOnOnSendRequest).toHaveBeenCalledWith("4");
  });

  it("Should dispatch the onClearAll props function when we choose the new button", () => {
    const radioButton = shallowWrapper.find("#radioButton");

    radioButton.simulate("change", {
      target: {
        value: "new",
      },
    });

    expect(spyOnClearAll).toHaveBeenCalled();
  });

  it("Should dispatch the handleUpdateAfterSave props function when we choose load button", () => {
    const radioButton = shallowWrapper.find("#radioButton");

    radioButton.simulate("change", {
      target: {
        value: "load",
      },
    });

    expect(spyOnHandleUpdateAfterSave).toHaveBeenCalled();
  });

  it("Should dispatch ondisplayStopSequence props function on select value in Auto-complete field ", () => {
    const AutoCompleteInput = shallowWrapper.find("#stopSequence_autoComplete");

    AutoCompleteInput.simulate("select", {
      target: {
        value: "St. Gallen to Zürich HB",
      },
    });

    expect(spyOnOndisplayStopSequence).toHaveBeenCalled();
  });
});
