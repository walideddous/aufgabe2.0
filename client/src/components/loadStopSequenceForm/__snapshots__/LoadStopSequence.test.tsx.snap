// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadStopSequence component Should match the LoadStopSequence component snapshot 1`] = `
<Card>
  <ForwardRef(InternalForm)
    form={
      Object {
        "__INTERNAL__": Object {
          "itemRef": [Function],
        },
        "getFieldError": [Function],
        "getFieldInstance": [Function],
        "getFieldValue": [Function],
        "getFieldsError": [Function],
        "getFieldsValue": [Function],
        "getInternalHooks": [Function],
        "isFieldTouched": [Function],
        "isFieldValidating": [Function],
        "isFieldsTouched": [Function],
        "isFieldsValidating": [Function],
        "resetFields": [Function],
        "scrollToField": [Function],
        "setFields": [Function],
        "setFieldsValue": [Function],
        "submit": [Function],
        "validateFields": [Function],
      }
    }
    layout="vertical"
  >
    <FormItem>
      <Memo([object Object])
        id="radioButton"
        onChange={[Function]}
        value="load"
      >
        <ForwardRef(RadioButton)
          id="load_button"
          value="load"
        >
          Load
        </ForwardRef(RadioButton)>
        <ForwardRef(RadioButton)
          id="new_button"
          value="new"
        >
          New
        </ForwardRef(RadioButton)>
      </Memo([object Object])>
    </FormItem>
    <FormItem
      label="Modes"
    >
      <ForwardRef(InternalSelect)
        id="mode_selector"
        onChange={[Function]}
        value="Choose mode"
      >
        <Option
          id="Choose mode"
          value="Choose mode"
        >
          Choose mode
        </Option>
        <Option
          id="13"
          value="13"
        >
          13
        </Option>
        <Option
          id="5"
          value="5"
        >
          5
        </Option>
        <Option
          id="8"
          value="8"
        >
          8
        </Option>
        <Option
          id="9"
          value="9"
        >
          9
        </Option>
        <Option
          id="2"
          value="2"
        >
          2
        </Option>
        <Option
          id="4"
          value="4"
        >
          4
        </Option>
      </ForwardRef(InternalSelect)>
    </FormItem>
    <FormItem
      label="Stop sequence name"
    >
      <ForwardRef(AutoComplete)
        allowClear={true}
        id="stopSequence_autoComplete"
        onChange={[Function]}
        onSelect={[Function]}
        placeholder="Enter stop sequence name"
        value=""
      >
        <Option
          key="c03295ea-5a3f-43e8-83ea-7736ce82cfd9"
          value="St. Gallen to Zürich HB"
        >
          <i
            className="fas fa-subway"
            style={
              Object {
                "color": "#1890ff",
                "margin": "0 10px",
              }
            }
          />
          St. Gallen to Zürich HB
        </Option>
      </ForwardRef(AutoComplete)>
    </FormItem>
  </ForwardRef(InternalForm)>
</Card>
`;
